cpp/open3d/t/geometry/TriangleMesh.cpp

```diff
@@ -1061,6 +1061,106 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
     return result;
 }
 
+// A helper to compute the vertex and triangle masks based on indices.
+// Additionally updates tris_cpu to new indices.
+template <typename intT>
+static void SBIUpdateMasksAndTrisCPUHelper(const core::Tensor &indices,
+                                           core::Tensor &vertex_mask,
+                                           core::Tensor &tris_mask,
+                                           core::Tensor &tris_cpu) {
+    const int64_t num_tris = tris_cpu.GetLength();
+    const int64_t num_verts = vertex_mask.GetLength();
+
+    // compute the vertices mask
+    intT *vertex_mask_ptr = vertex_mask.GetDataPtr<intT>();
+    const intT *indices_ptr = indices.GetDataPtr<intT>();
+    for (int64_t i = 0; i < indices.GetLength(); ++i) {
+        if (indices_ptr[i] >= num_verts) {
+            utility::LogError(
+                    "[SelectByIndex] indices contains index {} out of range. ",
+                    indices_ptr[i]);
+            continue;
+        }
+        vertex_mask_ptr[indices_ptr[i]] = 1;
+    }
+
+    // compute new vertix indices
+    std::vector<intT> prefix_sum(num_verts + 1, 0);
+    utility::InclusivePrefixSum(vertex_mask_ptr, vertex_mask_ptr + num_verts,
+                                &prefix_sum[1]);
+
+    // update the triangles with new indices and build the triangle mask
+    intT *tris_cpu_ptr = tris_cpu.GetDataPtr<intT>();
+    bool *tris_mask_ptr = tris_mask.GetDataPtr<bool>();
+    for (int64_t i = 0; i < num_tris; ++i) {
+        if (vertex_mask_ptr[tris_cpu_ptr[3 * i]] == 1 &&
+            vertex_mask_ptr[tris_cpu_ptr[3 * i + 1]] == 1 &&
+            vertex_mask_ptr[tris_cpu_ptr[3 * i + 2]] == 1) {
+            tris_cpu_ptr[3 * i] = prefix_sum[tris_cpu_ptr[3 * i]];
+            tris_cpu_ptr[3 * i + 1] = prefix_sum[tris_cpu_ptr[3 * i + 1]];
+            tris_cpu_ptr[3 * i + 2] = prefix_sum[tris_cpu_ptr[3 * i + 2]];
+            tris_mask_ptr[i] = true;
+        }
+    }
+}
+
+TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+    GetTriangleAttr().AssertSizeSynchronized();
+    GetVertexAttr().AssertSizeSynchronized();
+    if (GetTriangleIndices().GetDtype() == core::Int32) {
+        core::AssertTensorDtype(indices, core::Int32);
+    } else {
+        // we allow both Int32 and Int64 if the mesh indicies are Int64
+        core::AssertTensorDtypes(indices, {core::Int32, core::Int64});
+    }
+
+    // really copy triangles to CPU as we will modify the indices
+    core::Tensor tris_cpu =
+            GetTriangleIndices().To(core::Device(), true).Contiguous();
```

Comment:
The `true` argument here makes the "true" copy, which the `SelectFacesByMask` does not do. In my case, if I don't have it, the triangles do not match the original mesh if the device is CPU.